File: examples/claude-tool-use/invoice.py
import os
import httpx
import json
import anthropic
import uuid
from web3 import Web3
from eth_account.messages import encode_defunct
from dotenv import load_dotenv

load_dotenv(dotenv_path='../.env')

# Configuration
RPC_URL = os.getenv("RPC_URL", "https://mainnet.base.org")
USDC_ADDRESS = os.getenv("USDC_ADDRESS", "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913")
PRIVATE_KEY = os.getenv("PRIVATE_KEY")
API_BASE = os.getenv("API_BASE", "https://agentbureau-api.datafortress.cloud/v1")
ANTHROPIC_API_KEY = os.getenv("ANTHROPIC_API_KEY")

w3 = Web3(Web3.HTTPProvider(RPC_URL))
account = w3.eth.account.from_key(PRIVATE_KEY) if PRIVATE_KEY else None
client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)

def create_invoice_tool(customer_details, line_items):
    url = f"{API_BASE}/invoices"
    payload = {"customer_details": customer_details, "line_items": line_items}
    idempotency_key = str(uuid.uuid4())
    headers = {"Idempotency-Key": idempotency_key}
    
    # 1. Initial Call
    response = httpx.post(url, json=payload, headers=headers)
    
    if response.status_code == 402:
        data = response.json()
        intent_id = data.get("intent_id")
        
        payment_info = response.headers.get("PAYMENT-REQUIRED")
        parts = [p.strip() for p in payment_info.split(';')]
        amount_raw = int(float(parts[0]) * 1_000_000)
        wallet_address = parts[3]
        
        # 2. Send USDC
        usdc_abi = [{"constant":False,"inputs":[{"name":"_to","type":"address"},{"name":"_value","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"type":"function"}]
        usdc_contract = w3.eth.contract(address=USDC_ADDRESS, abi=usdc_abi)
        
        nonce = w3.eth.get_transaction_count(account.address)
        tx = usdc_contract.functions.transfer(wallet_address, amount_raw).build_transaction({
            'chainId': 8453,
            'gas': 100000,
            'gasPrice': w3.eth.gas_price,
            'nonce': nonce,
        })
        
        signed_tx = w3.eth.account.sign_transaction(tx, PRIVATE_KEY)
        tx_hash = w3.eth.send_raw_transaction(signed_tx.raw_transaction).hex()
        
        # 3. Wait and Sign
        w3.eth.wait_for_transaction_receipt(tx_hash)
        
        message = encode_defunct(text=f"AgentBureau Intent: {intent_id}")
        signed_message = w3.eth.account.sign_message(message, private_key=PRIVATE_KEY)
        signature = signed_message.signature.hex()
        
        # 4. Retry with payment signature and authorization
        retry_headers = {
            "Idempotency-Key": idempotency_key,
            "PAYMENT-SIGNATURE": tx_hash,
            "PAYMENT-AUTHORIZATION": signature
        }
        final_response = httpx.post(url, json=payload, headers=retry_headers)
        return final_response.json()
    
    return response.json()

def run_claude_agent():
    tools = [
        {
            "name": "create_invoice",
            "description": "Generates a professional invoice via AgentBureau.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "customer_details": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "email": {"type": "string"},
                            "address": {"type": "string"}
                        },
                        "required": ["name", "email", "address"]
                    },
                    "line_items": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "description": {"type": "string"},
                                "quantity": {"type": "number"},
                                "unit_price": {"type": "number"}
                            },
                            "required": ["description", "quantity", "unit_price"]
                        }
                    }
                },
                "required": ["customer_details", "line_items"]
            }
        }
    ]

    prompt = "Create an invoice for Acme Corp (billing@acme.com, 123 Business Way, London) for 10 units of AI Consulting at 150.00 each."

    response = client.messages.create(
        model="claude-sonnet-4-6",
        max_tokens=1024,
        tools=tools,
        messages=[{"role": "user", "content": prompt}]
    )

    if response.stop_reason == "tool_use":
        tool_use = response.content[-1]
        if tool_use.name == "create_invoice":
            result = create_invoice_tool(**tool_use.input)
            
            # Send result back to Claude
            final_response = client.messages.create(
                model="claude-sonnet-4-6",
                max_tokens=1024,
                tools=tools,
                messages=[
                    {"role": "user", "content": prompt},
                    {"role": "assistant", "content": response.content},
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "tool_result",
                                "tool_use_id": tool_use.id,
                                "content": json.dumps(result),
                            }
                        ],
                    },
                ]
            )
            print(final_response.content[0].text)

if __name__ == "__main__":
    run_claude_agent()

File: examples/typescript/invoice.ts
import { createWalletClient, http, parseAbi, publicActions } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { base } from 'viem/chains';
import * as dotenv from 'dotenv';
import { randomUUID } from 'crypto';

dotenv.config({ path: '../.env' });

// Configuration
const PRIVATE_KEY = (process.env.PRIVATE_KEY || '0x...') as `0x${string}`;
const RPC_URL = process.env.RPC_URL || 'https://mainnet.base.org';
const API_BASE = process.env.API_BASE || 'https://agentbureau-api.datafortress.cloud/v1';
const USDC_ADDRESS = (process.env.USDC_ADDRESS || '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913') as `0x${string}`;
const SERVICE_ENDPOINT = `${API_BASE}/invoices`;

const account = privateKeyToAccount(PRIVATE_KEY);
const client = createWalletClient({
  account,
  chain: base,
  transport: http(RPC_URL),
}).extend(publicActions);

async function runFlow() {
  const payload = {
    customer_details: {
        name: "Max Mustermann GmbH",
        email: "billing@example.com",
        address: "Hauptstr. 1, 10115 Berlin, Germany"
    },
    line_items: [
        {
            description: "AI Consulting Services",
            quantity: 10,
            unit_price: 150.00
        }
    ]
  };

  const idempotencyKey = randomUUID();
  console.log(`Calling ${SERVICE_ENDPOINT} with Idempotency-Key: {idempotencyKey}...`);

  // 1. Initial Call (Expect 402)
  const response = await fetch(SERVICE_ENDPOINT, {
    method: 'POST',
    headers: { 
      'Content-Type': 'application/json',
      'Idempotency-Key': idempotencyKey
    },
    body: JSON.stringify(payload),
  });

  if (response.status === 402) {
    const data = await response.json();
    const intentId = data.intent_id as `0x${string}`;
    
    const paymentRequired = response.headers.get('PAYMENT-REQUIRED');
    if (!paymentRequired) throw new Error('Missing PAYMENT-REQUIRED header');
    
    const parts = paymentRequired.split(';').map(s => s.trim());
    const amountRaw = BigInt(Math.floor(parseFloat(parts[0]) * 1_000_000));
    const receiver = parts[3] as `0x${string}`;

    console.log(`402 Received. Intent ID: ${intentId}`);
    console.log(`Payment Required: ${parts[0]} USDC to {receiver}`);

    // 2. Send USDC Transfer
    const usdcAbi = parseAbi([
      'function transfer(address to, uint256 amount) returns (bool)'
    ]);
    
    const hash = await client.writeContract({
      address: USDC_ADDRESS,
      abi: usdcAbi,
      functionName: 'transfer',
      args: [receiver, amountRaw],
    });

    console.log(`Transaction sent: ${hash}. Waiting for confirmation...`);
    await client.waitForTransactionReceipt({ hash });

    // 3. Sign the Intent ID for Agent Authorization
    console.log('Signing intent for authorization...');
    const authorization = await client.signMessage({
      message: `AgentBureau Intent: ${intentId}`,
    });

    // 4. Retry with payment signature and authorization
    console.log('Retrying with payment signature and authorization...');
    const finalResponse = await fetch(SERVICE_ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Idempotency-Key': idempotencyKey,
        'PAYMENT-SIGNATURE': hash,
        'PAYMENT-AUTHORIZATION': authorization,
      },
      body: JSON.stringify(payload),
    });

    if (finalResponse.ok) {
        console.log('Success!');
        console.log(await finalResponse.json());
    } else {
        console.log(`Final call failed with status ${finalResponse.status}`);
        console.log(await finalResponse.text());
    }
  } else {
    console.log(`Status: ${response.status}`);
    console.log(await response.text());
  }
}

runFlow().catch(console.error);
